src/main/kotlin/org/opensearch/replication/action/stop/TransportStopIndexReplicationAction.kt

```diff
@@ -140,10 +151,57 @@ class TransportStopIndexReplicationAction @Inject constructor(transportService:
             } catch (e: Exception) {
                 log.error("Stop replication failed for index[${request.indexName}] with error ${e.stackTraceToString()}")
                 listener.onFailure(e)
+            }finally {
+                enforceReplicationTasksRemoved(request)
             }
         }
     }
 
+    private fun enforceReplicationTasksRemoved(request: StopIndexReplicationRequest) {
+        val replicationTaskList: MutableList<String> = ArrayList()
+        for(retryCount in 1..3){
+            sleep(20000)
+            replicationTaskList.clear()
+            val tasks = clusterService.state().metadata().custom<PersistentTasksCustomMetadata>(PersistentTasksCustomMetadata.TYPE)
+            tasks.tasks().forEach {
+                if(isReplicationTask(it, request))
+                    replicationTaskList.add(it.id)
+            }
+            if (replicationTaskList.isEmpty()){
+                return
+            }
+        }
+        removeStaleReplicationTasksFromClusterState(request)
+    }
+
+    private  fun removeStaleReplicationTasksFromClusterState(request: StopIndexReplicationRequest) {
+        try {
+
+            val allTasks: PersistentTasksCustomMetadata =
+                clusterService.state().metadata().custom(PersistentTasksCustomMetadata.TYPE)
+
+            for (singleTask in allTasks.tasks()) {
+                if (isReplicationTask(singleTask, request)){
```

Comment:
Instead can we directly check the task status not being in assigned state and trigger removal?

Reply:
Removed logic to wait as we are now directly removing tasks which are not in assigned state.